hail/python/hail/ggplot/geoms.py

```diff
@@ -173,21 +173,32 @@ def apply_to_fig(self, grouped_data, fig_so_far: go.Figure, precomputed, facet_r
         legends = {}
         for df in grouped_data:
             values = self._get_aes_values(df)
-            trace_category = "trace1"
+            trace_categories = []
             for aes_name in self.aes_legend_groups:
                 category = self._get_aes_value(df, f"{aes_name}_legend")
-                trace_category = category if category is not None else trace_category
```

Comment:
When I tried to reproduce this issue locally, I was initially confused because the code example above (distilled from one of Dan's examples in the linked Zulip thread) produced the correct graph on my machine. However, another example that had worked fine for Dan evinced the issue for me. The example in question switched `shape` to be the aesthetic that had only one category (`"constant"`).

The reason for this is that this `for` loop is iterating through a `set`, not a `list`, so it does not have a deterministic order, and the root cause of the bug is that, in this line, if the case is encountered where there are two different aesthetics that each have categories, whichever one was seen most recently is the category used for the legend. This went unnoticed in my testing because it only shows up when there's a column with exactly one value in it mapped to either the `color` or `shape` aesthetic, not when one of the two is unmapped or both are mapped to columns with multiple values in them.

My change addresses the issue by storing all categories that are seen in this loop, and waiting until we have checked which column has multiple values to flatten them down to a single relevant category. In the case where multiple columns have multiple values, the category label we generate by this process is ignored in favor of the dummy legend introduced in https://github.com/hail-is/hail/pull/12254.